### ISSUE TITLE
Restarting Kernel infintely

### ISSUE THREAD

**Issue by linussun** [imatlab/imatlab#59]
Hi,

Just tried new python 3.6 jupyter lab install (latest) and Matlab 2018b kernel install (.egg is 2018a however). In any case, I'm getting infinite restarts of the kernel with any notebook. 

I am able to start matlab from python prompt: 
from matlab.engine import start_matlab; start_matlab()
<matlab.engine.matlabengine.MatlabEngine object at 0x10b39bc18>

However when looking at output of Jupyter lab, any  notebook with a MATLAB kernel, outputs the following...

[I 16:58:06.979 LabApp] Kernel started: 4c2f8e57-56aa-4776-a345-fb9304f224d3
[I 16:58:21.980 LabApp] KernelRestarter: restarting kernel (1/5), keep random ports
[I 16:58:36.993 LabApp] KernelRestarter: restarting kernel (1/5), keep random ports

... and it just continues and locks any interactivity with any working kernel. Any suggestions on what might be causing the problem?

Thank you!

**Comment by anntzer** [comment]
I don't use jupyterlab myself, so no idea, sorry.

**Comment by linussun** [comment]
This also happens in Juypter notebook:

[W 19:51:56.927 NotebookApp] Notebook Dropbox/MATLAB/Jupyter/notebooks/190705.ipynb is not trusted
[I 19:51:57.571 NotebookApp] Kernel started: 156c0411-a639-4548-a3f7-a84714d18745
[I 19:52:09.572 NotebookApp] KernelRestarter: restarting kernel (1/5), keep random ports
[I 19:52:21.582 NotebookApp] KernelRestarter: restarting kernel (1/5), keep random ports
[I 19:52:36.590 NotebookApp] KernelRestarter: restarting kernel (1/5), keep random ports
[I 19:52:51.601 NotebookApp] KernelRestarter: restarting kernel (1/5), keep random ports
^C[I 19:52:57.641 NotebookApp] interrupted

**Comment by anntzer** [comment]
Hard to tell, as I cannot repro this locally :( Perhaps mention this on the notebook tracker?

**Comment by linussun** [comment]
After reinstalling the MATLAB engine, and pip installing imatlab as admin, it appears to be working again without infinite kernel crashing/looping. Unclear why this occurs, but seems to happen every once in a while and needs a reinstall to fix. 

Python3.7/MATLAB2019a/Win10x64

**Comment by linussun** [comment]
Raising this again (now for Mac OSX). Kernel isn't working with fresh conda environment builds, getting:

[I 10:44:00.602 LabApp] Kernel started: b28e09df-0d75-4741-9908-67120e875197
[I 10:44:12.600 LabApp] KernelRestarter: restarting kernel (1/5), keep random ports
[I 10:44:24.608 LabApp] KernelRestarter: restarting kernel (1/5), keep random ports
and continues to restart. 

Jupyter Notebook 6.0.1
Jupyter lab 1.0.2
Python 3.6
Matlab 2018b
On MacOSX 10.14.6 

Python kernel works without restarting.

Any suggestions on how to debug this? 

Thanks.

**Comment by linussun** [comment]
Okay, I'm pretty sure of the newer commits this year broke the kernel. When I installed an older commit from last year (last commit of 2018), the kernel crash and restart doesn't occur. What I did to install: python -mpip install git+git://github.com/imatlab/imatlab@7c5d095931

The older and less well maintained Calysto kernel also doesn't crash/reboot continuously.

**Comment by anntzer** [comment]
can you bisect (https://git-scm.com/docs/git-bisect) it?

**Comment by linussun** [comment]
Hi Anntzer, 

It's a bit perplexing, now installing from current master branch works. I had been manually reinstalling the matlab kernel (ls /Applications/MATLAB_R2018b.app/extern/engines/python/) uninstalling ('pip uninstall imatlab') and reinstalling ('pip install imatlab' then 'python -mimatlab install') and various deleting of /usr/local/share/jupyter/kernel/imatlab to no avail. Only after I reinstalled Calysto and then the older imatlab branch, is the problem 'fixed' itself. Now the kernel doesn't continuously restart. I can even uninstall and reinstall from the current imatlab branch, so I can't reproduce the restarting kernel problem at the moment. I'll have to probably wait a few weeks/months of use before the restarting comes up again. 

While it probably won't fully solve the issue, if you could update the installation so that it supports conda environments (e.g. install kernelspec in /Users/anntzer/anaconda3/envs/share/jupyter/kernels) rather than install in /usr/local/share/jupyter/kernels this may help to isolate kernel issues.

I'm wondering if uninstalling and reinstalling from older kernels updated or downgraded a dependency which was causing the kernel to restart.

**Comment by anntzer** [comment]
Installing is provided by ipykernel (via https://github.com/imatlab/imatlab/blob/master/lib/imatlab/_kernel.py#L48); I don't know how ipykernel plays with conda envs but you may want to investigate that.

**Comment by djoshea** [comment]
FWIW, this happens reasonably often for me when I've updated `juptyerlab` or one of its dependencies. After setting up a debug environment where I can hook into the `imatlab` kernel with PyCharm, the problem has been related to parsing the history file. I find that deleting my Matlab history in `~/.matlab/R2019b/history.m` and `~/.matlab/R2019b/History.xml` often fixes this issue (these are for Linux, swap in your version). I'm not sure what's causing it, but I do often get warnings from Matlab if I open the desktop environment about the history file being corrupted by concurrent access, and I think the XML parser in the kernel chokes on this corrupted history file.

**Comment by anntzer** [comment]
I also get the warning but there's a button "Accept history as is", clicking on it seems to make things work just fine...

**Comment by djoshea** [comment]
Agreed. I don't know if they are connected. I think that Matlab itself may occasionally leave the history file in bad shape, but I cannot explain why the xml parsing routine called by imatlab would crash specifically when updating the jupyter packages via conda. All I'm saying is that sometimes on updates, imatlab crashes on startup and restarts indefinitely, and that in my case, deleting the history has fixed the issue. If I can reproduce the issue with a History file sometime, I will send it along, but it happens rarely enough that once I found this fix I stopped digging.

Thanks!
